src/peft/tuners/lora/layer.py

```diff
@@ -151,6 +167,23 @@ def reset_lora_parameters(self, adapter_name):
             nn.init.zeros_(self.lora_embedding_A[adapter_name])
             nn.init.normal_(self.lora_embedding_B[adapter_name])
 
+    def loftq_init_parameters(self, adapter_name, weight, num_bits, num_iter):
+        qweight, lora_A, lora_B = loftq_init(weight, num_bits, self.r[adapter_name], num_iter)
+        if adapter_name in self.lora_A.keys():
+            self.lora_A[adapter_name].weight.data = lora_A
+            self.lora_B[adapter_name].weight.data = lora_B
+        if adapter_name in self.lora_embedding_A.keys():
+            self.lora_embedding_A[adapter_name].data = lora_A
+            self.lora_embedding_B[adapter_name].data = lora_B
+        # TODO: better deal with this instead of creating weight here
+        if getattr(self, "base_layer", None) is None:
+            self.weight = nn.Parameter(qweight.to(weight.dtype), requires_grad=False)
+        elif is_bnb_available():
```

Comment:
Similar question: Should the weights not already be quantized? Is this to move them to CPU? If yes, why is that needed?